packages/dom-adapters/src/FormattingAdapter/index.ts

```diff
@@ -171,9 +173,77 @@ export class FormattingAdapter {
         return;
       }
 
-      const inlineElement = tool.createWrapper(event.detail.data.data);
+      const inputContent = input.textContent;
 
-      surround(inlineElement, input, textRange);
+      const rangeStart = Math.max(0, textRange[0] - 1)
+      const rangeEnd = inputContent !== null ? Math.min(inputContent.length, textRange[1] + 1) : 0;
+
+      const affectedFragments = this.#model.getFragments(blockIndex, dataKey, rangeStart, rangeEnd);
+
+      const lowerBoundary = affectedFragments[0]?.range[0] ?? textRange[0];
+      let upperBoundary = textRange[1];
+
+      for (const fragment of affectedFragments) {
+        upperBoundary = Math.max(upperBoundary, fragment.range[1]);
+      }
+
+      this.#rerenderRange(input, lowerBoundary, upperBoundary, affectedFragments);
+    }
+  }
+
+  /**
+   * Apply formatting of all affected fragments to the range with boundaries
+   *
+   * @param input - input element to apply formatting to
+   * @param lowerBoundary - lower boundary of the range
+   * @param upperBoundary - upper boundary of the range
+   * @param affectedFragments - model fragments that are used to format the range
+   */
+  #rerenderRange(input: HTMLElement, lowerBoundary: number, upperBoundary: number, affectedFragments: InlineFragment[]): void {
+    const range = document.createRange();
+    const [startNode, startOffset] = getBoundaryPointByAbsoluteOffset(input, lowerBoundary);
+    const [endNode, endOffset] = getBoundaryPointByAbsoluteOffset(input, upperBoundary);
+
+    if (startOffset === 0) {
+      range.setStartBefore(expandRangeNodeBoundary(startNode));
+    } else {
+      range.setStart(startNode, startOffset);
+    }
+
+    if (endOffset === endNode.textContent!.length) {
+      range.setEndAfter(expandRangeNodeBoundary(endNode, true));
+    } else {
+      range.setEnd(endNode, endOffset);
+    }
+
+    const extractedContent = range.cloneContents();
+
+    /**
+     * Create temporary container to allow formatting of the extracted content
+     */
+    const template = document.createElement('template');
+
+    template.innerHTML = extractedContent.textContent!;
```

Comment:
Would be more efficient as no need to clone the DOM nodes

```suggestion
    /**
     * Create temporary container to allow formatting of the extracted content
     */
    const template = document.createElement('template');

    template.innerHTML = range.toString()!;
```